Not compatible with Node v12.5

This currently is not an urgent issue because I'll just downgrade my local node version, but I'm just letting you know.

```
❯ node --version
v12.5.0
```

```
> auth0-magic@2.5.0 install /Users/mike/Development/auth0/magic
> node-gyp rebuild

  CXX(target) Release/obj.target/addon/extcrypto/extcrypto.o
../extcrypto/extcrypto.cc:27:39: error: too few arguments to function call, expected 4, have 3
    cb->Call(Null(isolate), argc, argv);
    ~~~~~~~~                          ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:4148:3: note: 'Call' declared here
  V8_WARN_UNUSED_RESULT MaybeLocal<Value> Call(Local<Context> context,
  ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8config.h:351:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../extcrypto/extcrypto.cc:35:39: error: too few arguments to function call, expected 4, have 3
    cb->Call(Null(isolate), argc, argv);
    ~~~~~~~~                          ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:4148:3: note: 'Call' declared here
  V8_WARN_UNUSED_RESULT MaybeLocal<Value> Call(Local<Context> context,
  ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8config.h:351:31: note: expanded from macro 'V8_WARN_UNUSED_RESULT'
#define V8_WARN_UNUSED_RESULT __attribute__((warn_unused_result))
                              ^
../extcrypto/extcrypto.cc:48:49: warning: 'NewFromUtf8' is deprecated: Use maybe version [-Wdeprecated-declarations]
    if (!kg) { return eret(isolate, cb, String::NewFromUtf8(isolate, "Unable to generate key")); }
                                                ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:2890:10: note: 'NewFromUtf8' has been explicitly marked deprecated here
  static V8_DEPRECATED(
         ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8config.h:311:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../extcrypto/extcrypto.cc:61:40: warning: 'NewFromUtf8' is deprecated: Use maybe version [-Wdeprecated-declarations]
      return eret(isolate, cb, String::NewFromUtf8(isolate, "Unable to generate key"));
                                       ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:2890:10: note: 'NewFromUtf8' has been explicitly marked deprecated here
  static V8_DEPRECATED(
         ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8config.h:311:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../extcrypto/extcrypto.cc:70:34: warning: 'NewFromUtf8' is deprecated: Use maybe version [-Wdeprecated-declarations]
    Local<String> rval = String::NewFromUtf8(isolate, key);
                                 ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:2890:10: note: 'NewFromUtf8' has been explicitly marked deprecated here
  static V8_DEPRECATED(
         ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8config.h:311:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../extcrypto/extcrypto.cc:82:23: error: no matching constructor for initialization of 'String::Utf8Value'
    String::Utf8Value ipem(pem);
                      ^    ~~~
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:3002:5: note: candidate constructor not viable: no known conversion from 'Local<v8::String>' to 'const v8::String::Utf8Value' for 1st argument
    Utf8Value(const Utf8Value&) = delete;
    ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:2995:5: note: candidate constructor not viable: requires 2 arguments, but 1 was provided
    Utf8Value(Isolate* isolate, Local<v8::Value> obj);
    ^
../extcrypto/extcrypto.cc:99:40: warning: 'NewFromUtf8' is deprecated: Use maybe version [-Wdeprecated-declarations]
      return eret(isolate, cb, String::NewFromUtf8(isolate, "Unable to generate key"));
                                       ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:2890:10: note: 'NewFromUtf8' has been explicitly marked deprecated here
  static V8_DEPRECATED(
         ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8config.h:311:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../extcrypto/extcrypto.cc:107:34: warning: 'NewFromUtf8' is deprecated: Use maybe version [-Wdeprecated-declarations]
    Local<String> rval = String::NewFromUtf8(isolate, pkey);
                                 ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:2890:10: note: 'NewFromUtf8' has been explicitly marked deprecated here
  static V8_DEPRECATED(
         ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8config.h:311:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../extcrypto/extcrypto.cc:118:23: error: no matching constructor for initialization of 'String::Utf8Value'
    String::Utf8Value ipem(pem);
                      ^    ~~~
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:3002:5: note: candidate constructor not viable: no known conversion from 'Local<v8::String>' to 'const v8::String::Utf8Value' for 1st argument
    Utf8Value(const Utf8Value&) = delete;
    ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:2995:5: note: candidate constructor not viable: requires 2 arguments, but 1 was provided
    Utf8Value(Isolate* isolate, Local<v8::Value> obj);
    ^
../extcrypto/extcrypto.cc:135:40: warning: 'NewFromUtf8' is deprecated: Use maybe version [-Wdeprecated-declarations]
      return eret(isolate, cb, String::NewFromUtf8(isolate, "Unable to generate key"));
                                       ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:2890:10: note: 'NewFromUtf8' has been explicitly marked deprecated here
  static V8_DEPRECATED(
         ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8config.h:311:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
../extcrypto/extcrypto.cc:143:34: warning: 'NewFromUtf8' is deprecated: Use maybe version [-Wdeprecated-declarations]
    Local<String> rval = String::NewFromUtf8(isolate, pkey);
                                 ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8.h:2890:10: note: 'NewFromUtf8' has been explicitly marked deprecated here
  static V8_DEPRECATED(
         ^
/Users/mike/.cache/node-gyp/12.5.0/include/node/v8config.h:311:29: note: expanded from macro 'V8_DEPRECATED'
  declarator __attribute__((deprecated(message)))
                            ^
7 warnings and 4 errors generated.
make: *** [Release/obj.target/addon/extcrypto/extcrypto.o] Error 1
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/local/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:200:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:272:12)
gyp ERR! System Darwin 18.6.0
gyp ERR! command "/usr/local/Cellar/node/12.5.0/bin/node" "/usr/local/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /Users/mike/Development/auth0/magic
gyp ERR! node -v v12.5.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok 
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! auth0-magic@2.5.0 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the auth0-magic@2.5.0 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/mike/.cache/npm/_logs/2019-07-09T19_55_08_927Z-debug.log
```

--- a/.circleci/config.yml
+++ b/.circleci/config.yml
@@ -1,18 +1,7 @@
 version: 2.1
-workflows:
-  main:
-    jobs:
-      - unit-tests:
-          version: 8
-      - unit-tests:
-          version: 10
-jobs:
-  unit-tests:
-    parameters:
-      version:
-        type: integer
-    docker:
-      - image: circleci/node:<< parameters.version >>
+commands:
+  runtests:
+    description: "Install deps and run tests"
     steps:
       - checkout
       - run:
@@ -22,3 +11,27 @@ jobs:
           name: Run Tests
           command: npm test
           when: always
+
+workflows:
+  main:
+    jobs:
+      - unit-tests-node8
+      - unit-tests-node10
+      - unit-tests-node12
+
+jobs:
+  unit-tests-node8:
+    docker:
+      - image: circleci/node:8
+    steps:
+      - runtests
+  unit-tests-node10:
+    docker:
+      - image: circleci/node:10
+    steps:
+      - runtests
+  unit-tests-node12:
+    docker:
+      - image: circleci/node:12
+    steps:
+      - runtests

--- a/binding.gyp
+++ b/binding.gyp
@@ -2,7 +2,10 @@
   "targets": [
     {
       "target_name": "addon",
-      "sources": [ "./extcrypto/extcrypto.cc" ]
+      "sources": [ "./extcrypto/extcrypto.cc" ],
+      "include_dirs" : [
+        "<!(node -e \"require('nan')\")"
+      ]
     }
   ]
 }

--- a/extcrypto/extcrypto.cc
+++ b/extcrypto/extcrypto.cc
@@ -6,12 +6,19 @@
 #include <openssl/pem.h>
 #include <openssl/bio.h>
 #include <openssl/bn.h>
+#include <nan.h>
 
 namespace extcrypto {
 
+  using Nan::Callback;
+  using Nan::FunctionCallbackInfo;
+  using Nan::GetFunction;
+  using Nan::New;
+  using Nan::Set;
+  using Nan::Utf8String;
+
   using v8::Exception;
   using v8::Function;
-  using v8::FunctionCallbackInfo;
   using v8::Isolate;
   using v8::Local;
   using v8::Null;
@@ -23,20 +30,22 @@ namespace extcrypto {
   void ret(Isolate* isolate, Local<Function> cb, Local<String> statement) {
     const uint64_t argc = 2;
     Local<Value> argv[argc] = { Null(isolate), statement };
+    Nan::Callback callback(cb);
 
-    cb->Call(Null(isolate), argc, argv);
+    callback.Call(v8::Object::New(isolate), argc, argv);
   }
 
 
   void eret(Isolate* isolate, Local<Function> cb, Local<String> statement) {
     const uint64_t argc = 1;
     Local<Value> argv[argc] = { Exception::Error(statement) };
+    Nan::Callback callback(cb);
 
-    cb->Call(Null(isolate), argc, argv);
+    callback.Call(v8::Object::New(isolate), argc, argv);
   }
 
 
-  void keygen(const FunctionCallbackInfo<Value>& args) {
+  void keygen(const Nan::FunctionCallbackInfo<Value>& args) {
     Isolate* isolate   = args.GetIsolate();
     Local<Function> cb = Local<Function>::Cast(args[0]);
 
@@ -74,12 +83,12 @@ namespace extcrypto {
   }
 
 
-  void extract(const FunctionCallbackInfo<Value>& args) {
+  void extract(const Nan::FunctionCallbackInfo<Value>& args) {
     Isolate* isolate   = args.GetIsolate();
     Local<String> pem  = Local<String>::Cast(args[0]);
     Local<Function> cb = Local<Function>::Cast(args[1]);
 
-    String::Utf8Value ipem(pem);
+    Nan::Utf8String ipem(pem);
     char* skey(*ipem);
 
     BIO* bio = BIO_new(BIO_s_mem());
@@ -110,12 +119,12 @@ namespace extcrypto {
     return ret(isolate, cb, rval);
   }
 
-  void extractSPKI(const FunctionCallbackInfo<Value>& args) {
+  void extractSPKI(const Nan::FunctionCallbackInfo<Value>& args) {
     Isolate* isolate   = args.GetIsolate();
     Local<String> pem  = Local<String>::Cast(args[0]);
     Local<Function> cb = Local<Function>::Cast(args[1]);
 
-    String::Utf8Value ipem(pem);
+    Nan::Utf8String ipem(pem);
     char* skey(*ipem);
 
     BIO* bio = BIO_new(BIO_s_mem());
@@ -146,12 +155,26 @@ namespace extcrypto {
     return ret(isolate, cb, rval);
   }
 
-  void init(Local<Object> exports) {
-    NODE_SET_METHOD(exports, "keygen", keygen);
-    NODE_SET_METHOD(exports, "extract", extract);
-    NODE_SET_METHOD(exports, "extractSPKI", extractSPKI);
-  }
+  void init(v8::Local<v8::Object> exports) {
+    v8::Local<v8::Context> context = exports->CreationContext();
+    exports->Set(context,
+                 Nan::New("keygen").ToLocalChecked(),
+                 Nan::New<v8::FunctionTemplate>(keygen)
+                     ->GetFunction(context)
+                     .ToLocalChecked());
 
+    exports->Set(context,
+                 Nan::New("extract").ToLocalChecked(),
+                 Nan::New<v8::FunctionTemplate>(extract)
+                     ->GetFunction(context)
+                     .ToLocalChecked());
+
+    exports->Set(context,
+                 Nan::New("extractSPKI").ToLocalChecked(),
+                 Nan::New<v8::FunctionTemplate>(extractSPKI)
+                     ->GetFunction(context)
+                     .ToLocalChecked());
+  }
 
   NODE_MODULE(NODE_GYP_MODULE_NAME, init)
 }

--- a/package.json
+++ b/package.json
@@ -10,7 +10,8 @@
   },
   "dependencies": {
     "bcrypt": "3.0.6",
-    "libsodium-wrappers-sumo": "0.7.3"
+    "libsodium-wrappers-sumo": "0.7.3",
+    "nan": "^2.14.0"
   },
   "scripts": {
     "test": "./node_modules/.bin/mocha --delay ./test/magic.test.js ./test/magic.test.vectors.js"

